fix: filter monitors by tags added through monitor.use

devcorpio opened the pull request:
Fixes: https://github.com/elastic/synthetics/issues/775

# Context

the `--tags` flag in the push command was only taking into account tags passed as an argument in the `journey` function

```
journey({ name: 'My Example Journey', tags: ['foo']}, ({ page, params }) => {
   // ... 
})
```

# Fix

Tags added through `monitor.use` are now included in the journey before filtering monitors by `tag`.


## How to test

1. Create [project monitor](https://www.elastic.co/guide/en/observability/current/synthetics-get-started-project.html#synthetics-get-started-project-init)
2. Make sure you add the tag via `monitor.use`.  For instance:

  ```
journey('My Example Journey', ({ page, params }) => {
  monitor.use({
    tags: ['foo']
  });

  step('launch application', async () => {
    await page.goto(params.url);
  });

  step('assert title', async () => {
    const header = await page.locator('h1');
    expect(await header.textContent()).toBe('todos');
  });
});

  ```
 
3. execute `SYNTHETICS_API_KEY=YOUR_KEY npm run push`. You should see 4 added tests (at least if you stay with the default tests).
4. execute `SYNTHETICS_API_KEY=YOUR_KEY npm run push -- --tags foo`. You should see 3 removed tests and one unchanged

**Recording**:


https://github.com/elastic/synthetics/assets/15065076/7f1d3f60-0658-4257-9cee-1580a185522a

emilioalvap reviewed (approved):
LGTM

vigneshshanmugam commented on src/core/runner.ts:
```diff
@@ -410,6 +407,15 @@ export default class Runner {
+
+      /**
+       * Add the tags created through `monitor.use` to the journey
+       */
```
Why not add this as part of update monitor which handles all config from 'monitor.use'?

vigneshshanmugam commented on src/core/runner.ts:
```diff
@@ -410,6 +407,15 @@ export default class Runner {
+       * Add the tags created through `monitor.use` to the journey
+       */
+      journey.addTags(journey.monitor.config.tags);
+      if (!journey.isMatch(options.match, options.tags)) {
```
I don't think this is the correct behaviour. Monitor.use should never influenced the local runner. It should be only applicable to the pushed monitors.

vigneshshanmugam commented on __tests__/core/runner.test.ts:
```diff
@@ -809,4 +809,18 @@ describe('runner', () => {
     expect(monitors[0].config.name).toBe('j1');
   });
+
+  it('runner - build monitors filtered through "tags" added updating monitor config', async () => {
```
Instead of having this test in runner, it should be on the push command as monitor.use only touches the pushed monitors and never deals with runner parameters.

devcorpio replied:
thank you for the context 🚀 . I will have a look at this

devcorpio replied:
done!

andrewvc commented:
Have we tested this with the `synthetics.config.ts` and tags declared there?

devcorpio commented:
> Have we tested this with the synthetics.config.ts and tags declared there?

Thanks for asking this, @andrewvc. I didn't know it was possible to add tags to that file. This scenario wasn't working, I added the required changes for that, too. 

These are the commits related to that change:
- https://github.com/elastic/synthetics/pull/776/commits/854d5d84f3003af8441529f7ee22e16066e36abd
- https://github.com/elastic/synthetics/pull/776/commits/be0febadb8d900ddbfabc78dc89c9142b8e37eee

You will see I added a few tests on that area as well

devcorpio commented on src/options.ts:
```diff
@@ -115,6 +115,7 @@ export function normalizeOptions(cliArgs: CliArgs): RunOptions {
   options.locations = cliArgs.locations ?? monitor?.locations;
   options.privateLocations =
     cliArgs.privateLocations ?? monitor?.privateLocations;
+  options.tags = cliArgs.tags ?? monitor?.tags;
```
@vigneshshanmugam @andrewvc 

I have a question:

This block starts with:

 /**
   * Get the default monitor config from synthetics.config.ts file
  */

But I've noticed that we are just adding:

- schedule
- locations
- privateLocations
- tags (added in this PR)

If you have a look at https://github.com/elastic/synthetics/blob/main/src/dsl/monitor.ts#L47

you will see there are options that we are not getting there. Is there a reason for that? Probably I'm missing context, too. but I believe it's important to ask, just in case

devcorpio commented:
closing PR. Reason [here](https://github.com/elastic/synthetics/issues/775#issuecomment-1587849721)

devcorpio commented:
Closing PR. Reason [here](https://github.com/elastic/synthetics/issues/775#issuecomment-1587849721)
